scripts/gen-index.cts

```diff
@@ -0,0 +1,148 @@
+		this.path = parse(sourceFile.fileName);
+		this.isPrivate = this.path.name.startsWith('_');
+		this.exports = this.parseExports();
+		this.exportInclusions = this.deriveExportInclusions();
+	}
+
+	public generateExportSpecifiers(which: keyof ModuleExportNodes) {
+		return [...new Set(this.exports[which].map((node) => ts.getNameOfDeclaration(node)!.getText(this.sourceFile)!))].map((name) =>
+			ts.factory.createExportSpecifier(false, undefined, name)
+		);
+	}
+
+	/**
+	 * Rules:
+	 * 1. If the module begins with `_`, ignore it
+	 *    a) unless the module contains types, in which case those exclusively should be included
+	 * 2. If the module exclusively contains types, use `export type *`, otherwise use `export *`
+	 **/
+	public toExportDeclaration(packageDir: string): ts.ExportDeclaration | undefined {
+		const { useModule, useNormal, useTypes } = this.exportInclusions;
+		if (!useModule) return;
+
+		// TODO: make this more efficient
+		const typeExportSpecifiers = this.generateExportSpecifiers('types');
+		const exportSpecifiers = typeExportSpecifiers.concat(useNormal ? this.generateExportSpecifiers('normal') : []);
+		const relativePath = relative(packageDir, this.path.dir);
+		const adjustedPath = (this.path.base === 'index.ts' ? relativePath : join(relativePath, this.path.name)).split(winSep).join(posixSep);
+
+		return ts.factory.createExportDeclaration(
+			undefined,
+			!useNormal && useTypes,
+			this.isPrivate ? ts.factory.createNamedExports(exportSpecifiers) : undefined,
+			ts.factory.createStringLiteral(`./${adjustedPath}`, true),
+			undefined
+		);
+	}
+
+	private deriveExportInclusions(): ModuleExportInclusions {
+		const typesAvailable = this.exports.types.length > 0;
+
+		return {
+			useModule: !this.isPrivate || typesAvailable, // rule 1 & 1a
+			useNormal: !this.isPrivate && this.exports.normal.length > 0,
+			useTypes: typesAvailable
+		};
+	}
+
+	private parseExports(): ModuleExportNodes {
+		const normal: ts.Declaration[] = [];
+		const types: ts.Declaration[] = [];
+
+		// TODO: skip normals where this.isPrivate?
+		this.sourceFile.forEachChild((node) => {
+			if (!ts.isDeclarationStatement(node) || !isExported(node)) return;
+			isType(node) ? types.push(node) : normal.push(node);
+		});
+
+		return {
+			normal,
+			types
+		};
+	}
+}
+
+async function findIndexOrModules(dir: string, depth: number = 0): Promise<string[]> {
+	const contents = await readdir(dir, { withFileTypes: true });
+	let results: string[] = [];
+	for (const item of contents) {
+		if (item.name === 'common') continue;
+		const itemPath = join(dir, item.name);
+		if (item.isFile()) {
+			if (!item.name.endsWith('.ts')) continue;
+			if (item.name === 'index.ts') {
+				// TODO: process index to support re-exports (see #750)
+				if (depth === 0) continue;
+				results = [itemPath];
+				break;
+			}
+			results.push(itemPath);
+			// TODO: there are likely more efficient ways to determine if a directory has an auxiliary index
```

Comment:
Resolve in this PR, or delete the TODO. Your call. Adding a random TODO is poor management of technical debt. Taking up tech. debt is something we only do if there is a good reason to do so.